### ISSUE TITLE
error: invalid register `tp`: the thread pointer cannot be used as an operand for inline asm

### ISSUE THREAD

**Issue by vsukhoml** [rust-lang/rust#106947]
There are certain assumptions made by Rust (LLVM?) about register usage on various platforms (RISC-V in this example) which are true only for some platforms, but not true for variety of embedded use cases.  `clang` provide some workarounds with register variables extension and options like `-ffixed-x10`, but in Rust it is not clear how to efficiently get same functionality without copy of 'reserved' register into 'general purpose one'.

```
 unsafe fn get_sys_buf() -> *mut u8 {
        let syscall_buf: *mut u8;
        unsafe { asm!("", out("tp") syscall_buf) };
        syscall_buf
    }
```

**Comment by workingjubilee** [comment]
> There are certain assumptions made by Rust (LLVM?) about register usage on various platforms (RISC-V in this example) which are true only for some platforms, but not true for variety of embedded use cases. `clang` provide some workarounds with register variables extension and options like `-ffixed-x10`, but in Rust it is not clear how to efficiently get same functionality without copy of 'reserved' register into 'general purpose one'.

It's not always clear what LLVM reserves for itself or only reserves for itself on Tuesday and Thursday, platform-wise. If there's documentation in LLVM you can point to that justifies the workarounds clang is offering, that could help greatly in improving the situation, relatively speaking.

**Comment by vsukhoml** [comment]
Oh, I got error message: 
`error: invalid register tp: the thread pointer cannot be used as an operand for inline asm` for the code` for the code above.
See https://godbolt.org/z/v187f6P8v

`tp` is a thread-pointer on RISC-V according to https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md, so normally (Linux I guess?) it is used to address TLS. But for embedded code I need to perform certain operations manually. In C I'd use `register void *syscall_buf asm("tp");`  as in https://godbolt.org/z/5azbqzjo7 

I'm not sure where exactly this error is originated since LLVM allows reads from `tp`. Can it be from https://docs.rs/rustc-ap-rustc_target/721.0.0/src/rustc_ap_rustc_target/asm/riscv.rs.html#131?

**Comment by workingjubilee** [comment]
Hmm. So I took a look at this and this sequence works ([Godbolt](https://godbolt.org/z/Wdzn3fT3T))
```rust
#![no_std]
#![deny(unsafe_op_in_unsafe_fn)]

use core::arch::asm;
pub unsafe fn get_sys_buf() -> *mut u8 {
    let syscall_buf: *mut u8;
    unsafe { asm!("mv {syscall_buff}, tp", syscall_buff = out(reg) syscall_buf) };
    syscall_buf
}
```

Is this a satisfactory alternative? I can see how this has problems in terms of documentation, at least... it's not necessarily obvious that such is what you should do.

**Comment by vsukhoml** [comment]
Sure, this one works, but it takes 1 register, and 1 `mv` instruction, which is not desirable for size-optimized embedded code.  When using `tp` directly it is possible to use `tp`-relative addressing, free one register for other purposes.

I'm not sure why this restriction (not allow `gp` and `tp`) exists and wherever it make sense to make it configurable. Sure, it came from interpretation of RISC-V ABI, but this ABI is for Rust code and in inline asm it should be possible to use any registers, especially as output.

**Comment by workingjubilee** [comment]
Hm. LLVM may make certain assumptions about how a register is used, such as assuming human users will never clobber or inspect it, to optimize the compiler assembly around your inline assembly, such as eliding prologs and epilogs, deferring fixing up the stack (and then doing so correctly), etc.

Which is relevant here since it's about what you want, it sounds like. That is, it seems what you want is specifically to coax LLVM into tp-relative addressing for operations in the larger function where this smaller one gets inlined, without using an instruction at all. It may be helpful if you can provide a more "worked" example that uses this smaller function in a very slightly larger function that you'd expect to compile to tp-relative addressing loads and stores. I can create my own example, obviously, but it would remove any doubt about the target.

As I understand the compiler code for Rust's "inline assembler", we make no distinction during compilation between reading and writing for an operand. So one possible change might be to introduce a distinction between whether an operand can be used with `in` vs. `inout` vs. `out`, and so on.

**Comment by vsukhoml** [comment]
It seems this specific assumption is not enforced by LLVM, but by Rust compiler and at the same is valid only for certain targets like -linux. When compiled for something like `riscv32-unknown-unknown-elf` - common target for firmware with custom OS it is no longer valid. Yes, compiler don't use `tp` explicitly in many cases, but it doesn't me OS/apps can't.

In my particular case I want to use `tp` without modifying it, as a pure output register and use it for addressing. LLVM and gcc are fine with that in C when I use register variables. Yes, it may be Rust compiler should start distinguish input/output for these `special` registers or treat it differently depending on target triple.

For example in https://godbolt.org/z/67h5fercG I get
```
example::get_b:
        mv      a0, tp
        lw      a0, 4(a0)
        ret
```

instead of:
```
example::get_b:
        lw      a0, 4(tp)
        ret
```
and equivalent code in C https://godbolt.org/z/exe1afvPs which produce what I want, though with slightly different means as gcc/clang's inline asm for RISC-V doesn't allow explicit register name (x86 allows this for certain registers).

**Comment by Amanieu** [comment]
The ability to read a global register and have it used in-place by the compiler is outside the scope of `asm!`, and if you really need it then a new interface needs to be provided to expose this feature.

If you look at the LLVM IR that is generated by Clang for your C example, you will see that it uses the 
`llvm.read_register.i32` intrinsic, and not inline assembly.

**Comment by vsukhoml** [comment]
Is it possible to use this intrinsic in Rust via `#[link_name = "llvm.read_register.i32"] ? I tried https://godbolt.org/z/Td1Y4f3W9, but apparently need to know argument types which aren't properly documented.

**Comment by Amanieu** [comment]
The intrinsic takes a `metadata` argument in LLVM IR, which is not something that can be directly created from Rust code.

**Comment by vsukhoml** [comment]
So, it is impossible for now. Also looks like `clang` accepts that syntax but doesn't read from `tp` in the https://godbolt.org/z/cxoY3bYW8 (gcc works correctly).

However, restriction for asm!() in Rust is Rust specific in https://docs.rs/rustc-ap-rustc_target/721.0.0/src/rustc_ap_rustc_target/asm/riscv.rs.html#131 and can be only applied for certain targets like windows or linux, but dropped for `unknown`  which is used for embedded.

**Comment by Amanieu** [comment]
No, the restriction cannot be dropped because LLVM always treats `tp` as a reserved register on all platforms. If it were not reserved then it would mean that LLVM could freely use `tp` as a general-purpose register in a function, which is clearly not the case in your example.

**Comment by vsukhoml** [comment]
I see, thanks. I was about to rebuild Rust without this restriction to check how it works. 
And I might be ok with custom ABIs though - like ability to return values in 4 registers vs. 2 on RISC-V, etc.
